network/src/peer_store/addr_manager.rs

```diff
@@ -0,0 +1,96 @@
+use crate::peer_store::types::{AddrInfo, IpPort};
+use rand::Rng;
+use std::collections::{HashMap, HashSet};
+
+#[derive(Default)]
+pub struct AddrManager {
```

Comment:
IMO our code is not such complex, I'd rather do not to introduce another unnecessary dependency.

Reply:
And the intention here is to arrange id for each `AddrInfo`, so we can use multiple way's to accces it, such as `random_ids`. 
It's not just for the sequence access, when we delete a elem from `IndexMap` we need to depent the internal implement to keep the `random_ids` consistent with `IndexMap` internal.

Reply:
You can access `IndexMap` via index, the original requirement of `&mut self` in a getter fn `get_xxx` is weird.  The `get_random` function can be simplified to:
```rust
pub fn get_random<F>(&self, count: usize, filter: F) -> Vec<AddrInfo> {
    let mut indexes = (0..addres.len()).collect();
    thread_rng().shuffle(&mut indexes);
    indexes.iter().for_each(|i| {
        let (ip_port, addr_info) = self.addres.get_index(i);
        ...
    })
}
```

btw, there is a bug in `get_random` fn, result limit `count` param is ignored, it will return all addresses, please add an unit test and fix it.

Reply:
Fix the `count` and `serde`.

This change may reduce a few lines of code, but it makes code slower. The current implementation only exchange `count` number elements, but with `shuffle` we need to exchange `addrs.len()` number elements. 

I do not think the `get` with `&mut self` is unacceptable, in `LRUCache` we already use it. I'am ok to change the name `get_xxx` to `fetch_xxx`.